### PR TITLE
specify amount in 'approve' method

### PR DESCRIPTION
**Pull request by m-cote**
Fixes #SB-363

## Description of the changes

- adds optional param `amount` to method `approve` that allows to revoke approval by setting its value to 0 or set allowance to an arbitrary amount of tokens

### CODE REVIEW

File: src/bridge/trx/index.ts [hunk 1]
```diff
@@ -126,25 +126,27 @@ export class TronBridge extends Bridge {
     );
   }
 
-  async approve(approveData: ApproveData): Promise<TransactionResponse> {
-    const { tokenAddress, spender, owner } = approveData;
+  async approve(params: ApproveParamsDto): Promise<TransactionResponse> {
+    const { tokenAddress, spender, owner, amount } = params;
     const tokenContract = await this.getContract(tokenAddress);
+    const amountHex = amount === undefined ? MAX_AMOUNT : amountToHex(amount);
```
Comment (YaroslavNekryach) [inline src/bridge/trx/index.ts:132]: This is the only case when == is better than === If you use == undefined it false on null as well

File: src/bridge/trx/index.ts [hunk 2]
```diff
@@ -126,25 +126,27 @@ export class TronBridge extends Bridge {
...
     const transactionHash = await tokenContract
```
Comment (YaroslavNekryach) [inline src/bridge/trx/index.ts:133]: Try to use buildRawTransactionApprove here

File: src/bridge/trx/index.ts [hunk 3]
```diff
@@ -126,25 +126,27 @@ export class TronBridge extends Bridge {
...
-      .approve(spender, MAX_AMOUNT)
+      .approve(spender, amountHex)
       .send({ from: owner });
     /* eslint-disable-next-line @typescript-eslint/no-unsafe-argument */
```
Comment (YaroslavNekryach) [inline src/bridge/trx/index.ts:136]: remove ignore

File: src/bridge/index.ts [hunk 4]
```diff
@@ -135,4 +141,34 @@ export class BridgeService {
       ),
     };
   }
+
+  private async prepareApproveParams(
```
Comment (YaroslavNekryach) [inline src/bridge/index.ts:145]: If you use int amount, this method wil be simpler

**Review by YaroslavNekryach — CHANGES_REQUESTED** [review]

File: src/bridge/trx/index.ts (hunk 3 above)
Comment (m-cote) [reply src/bridge/trx/index.ts]: I disabled check `@typescript-eslint/no-unsafe-argument` in .eslintrc.json and removed this comment

**Review by Kozer4 — APPROVED** [review]

File: src/bridge/index.ts (hunk 4 above)
Comment (m-cote) [reply src/bridge/index.ts]: I changed `approve` method to accept an integer amount

File: src/bridge/evm/index.ts [hunk 5]
```diff
@@ -129,13 +129,13 @@ export class EvmBridge extends Bridge {
...
+  async approve(params: ApproveParamsDto): Promise<TransactionResponse> {
+    const { tokenAddress, spender, owner, amount } = params;
     const tokenContract = this.getContract(erc20abi as AbiItem[], tokenAddress);
 
     const approveMethod = await tokenContract.methods.approve(
       spender,
-      MAX_AMOUNT
+      amount === undefined ? MAX_AMOUNT : amountToHex(amount)
     );
     const estimateGas = await approveMethod.estimateGas({ from: owner });
```
[the remark this answers is not in the document]
Comment (m-cote) [reply src/bridge/evm/index.ts]: I replaced custom send transaction logic with sending transaction with buildRawTransactionApprove method.

File: src/bridge/trx/index.ts (hunk 2 above)
Comment (m-cote) [reply src/bridge/trx/index.ts]: I replaced custom send transaction logic with sending transaction with buildRawTransactionApprove method.

**Pull request merged**